### ISSUE TITLE
Support parsing of Flow annotations

### ISSUE THREAD

**Issue by agigleux** [SonarSource/sonar-javascript#584]
Migrated from: https://jira.sonarsource.com/browse/SONARJS-828
Flow: https://flow.org/

**Comment by sattaman** [comment]
+1

**Comment by ummahusla** [comment]
+1

**Comment by trietphamnm** [comment]
+1

**Comment by himynameisjon** [comment]
+1

**Comment by r085741** [comment]
+1

**Comment by bendersej** [comment]
+1

**Comment by memelet** [comment]
Can sonarjs be used with flow at all? We have a bunch of project that employ flow. Do people configure sonar to use js output from flow maybe?

**Comment by inverno** [comment]
@memelet what you propose could work and with source maps it might even be vaguely useful, but do consider that the reason we are not yet supporting flow annotation is not related to parsing them without failing, it's related to the information content they provide : if you are using flow and we run analysis on your sources (or output) without understanding the information provided by flow we might raise a lot of issues that are not relevant.
In short : you might try that, but please don't be upset if the analyzer raises `obviously wrong` issues.

**Comment by kesor** [comment]
@inverno at the moment the scanner cannot parse files annotated with flow. Any files that have flow annotations cause an `ERROR: Unable to parse file` and are being ignored.

**Comment by cszczotka** [comment]
hi,
Do you have plan to implement support for flow syntax in sonarqube javascript plugin ? Maybe you have same a workaround how to scan javascript code where is used flow ?
Regards,
Czeslaw

**Comment by pawelk** [comment]
@cszczotka  I have added [flow-remove-types](https://www.npmjs.com/package/flow-remove-types) to my project and pointed Sonarqube to scan the non-annotated JS. It's far from ideal and meant to be temporary, but at least the code gets scanned.

**Comment by vilchik-elena** [comment]
Flow features:
- [ ] typed variable declaration
- [ ] function declarations (declaration, expression, arrow, method): types for parameters, type for return, optional parameters, type for rest parameter
- [ ] optional type
- [ ] literal type
- [ ] function type, with generics
- [ ] object type, optional properties in object type, exact object type, indexer property for object type (aka map object)
- [ ] array types
- [ ] parenthesis in type (e.g. `let arr2: (?number)[] = [1, 2];`)
- [ ] tuple types
- [ ] typed properties in class declaration
- [ ] class generics (in class declaration and when using this class in variable declaration)
- [ ] type alias statement, with `opaque`, with supertype
- [ ] import statements
- [ ] export statements
- [ ] declare type statement
- [ ] interface type (properties, method, modifiers `+` and `-`)
- [ ] generics: function declarations, class declarations, interface declarations, type aliases, function type.[ Parameterized generics](https://flow.org/en/docs/types/generics/#toc-parameterized-generics). Super type for generic parameter ([link](https://flow.org/en/docs/types/generics/#toc-adding-types-to-generics)).
- [ ] union types
- [ ] [disjoint Unions](https://flow.org/en/docs/types/unions/#toc-disjoint-unions)
- [ ] intersection types
- [ ] typeof operator
- [ ] type-casting
- [ ] Utility Types ($Keys, $Diff etc)

**Comment by vilchik-elena** [comment]
**Some Links**
- parser from babel https://github.com/babel/babylon/blob/master/src/plugins/flow.js
- [flow online (from babel) ](http://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=MYewdgzgLgBAhgLhmArgWwEYFMBOMC8MAzANwBQZAbnHgN4AeANDAJ4C-BMDSATMy0iIcKoSCAA2WAHTiQAcwAU9AJRA&debug=false&circleciRepo=&evaluate=false&lineWrap=false&presets=es2015%2Creact%2Cstage-2&prettier=false&targets=&version=6.26.0)

**Comment by vilchik-elena** [comment]
There is still some syntax not supported. See this issue #771

**Comment by mkruneva** [comment]
Hi, I am still having issues with flow syntax ? Is there a workaround or any updates? 

Thank you

**Comment by vilchik-elena** [comment]
@mkruneva comment on #771 with examples
